src/block/stdio_executor.rs

```diff
@@ -192,45 +318,188 @@ impl<B: Backend> StdIoBackend<B> {
         match request_type {
             RequestType::In => {
                 self.check_access(total_len / SECTOR_SIZE, request.sector())?;
+                // Total data length should fit in an u32 for further writing in the used ring.
+                if total_len > u32::MAX as u64 {
+                    return Err(Error::InvalidDataLength);
+                }
                 for (data_addr, data_len) in request.data() {
                     mem.read_exact_from(*data_addr, &mut self.inner, *data_len as usize)
-                        .map_err(Error::Read)?;
-                    bytes_from_dev += data_len;
+                        .map_err(|e| {
+                            match e {
+                                GuestMemoryError::PartialBuffer {
+                                    completed,
+                                    expected: _,
+                                } =>
+                                // The `as u32` cast is safe, since completed < data_len (which is
+                                // an u32).
+                                {
+                                    bytes_to_mem += completed as u32
+                                }
+                                _ => {}
+                            }
+                            Error::Read(e, bytes_to_mem)
+                        })?;
+                    // This can not overflow since we checked right before the loop that `total_len`
+                    // fits in an u32.
+                    bytes_to_mem += data_len;
                 }
             }
             RequestType::Out => {
                 self.check_access(total_len / SECTOR_SIZE, request.sector())?;
-                let mut bytes_to_dev = 0;
                 for (data_addr, data_len) in request.data() {
                     mem.write_all_to(*data_addr, &mut self.inner, *data_len as usize)
                         .map_err(Error::Write)?;
-                    bytes_to_dev += data_len;
                 }
             }
             RequestType::Flush => return self.inner.fsync().map(|_| 0).map_err(Error::Flush),
+            RequestType::GetDeviceID => {
+                let device_id = self
+                    .device_id
+                    .ok_or(Error::Unsupported(VIRTIO_BLK_T_GET_ID))?;
+                // The length of data MUST be VIRTIO_BLK_ID_BYTES bytes for VIRTIO_BLK_T_GET_ID
+                // requests.
+                if total_len != VIRTIO_BLK_ID_BYTES as u64 {
+                    return Err(Error::InvalidDataLength);
+                }
+                for (data_addr, data_len) in request.data() {
+                    // The device_id accesses are safe because we checked that the total data length
+                    // is VIRTIO_BLK_ID_BYTES, which is the size of the id as well.
+                    mem.read_exact_from(
+                        *data_addr,
+                        &mut &device_id[bytes_to_mem as usize..(*data_len + bytes_to_mem) as usize],
```

Comment:
I used a cursor here but a slice range should be also fine: https://github.com/rust-vmm/vm-virtio/pull/56/files#diff-ae9f3a857e7d373fd75ef4481f3d920ae3416de07e3a67c43fe8241443cf5fdfR220
The error handling looks more complex. I have not looked into it if that is necessary but probably you di.

Reply:
I am exaggerating a bit here indeed. It is probably impossible to have a partial read when the max size of access is 20 bytes, so we could probably remove the match below. My preference would be to keep the check though.